### DISCUSSION TITLE
shutters and hue emulation support

### DISCUSSION THREAD

**Discussion by oponyx** [arendst/Tasmota#13509] [Ideas]
Hi everybody,
I'm trying to configure Tasmota to drive the shutters in percentage mode using hue emulation and Alexa but don't find a way..
I don't figure how to make the shutter visible as dimmer to Alexa.

Digging the code i see inside the file xdrv_20_hue.ino at line 879:
    if (change) {
#ifdef USE_SHUTTER
      if (ShutterState(device)) {
        AddLog(LOG_LEVEL_DEBUG, PSTR("Settings->shutter_invert: %d"), Settings->shutter_options[device-1] & 1);
        ShutterSetPosition(device, bri * 100.0f );
      } else
#endif

then it seems it's supposed to set the shutter position at the desired bri value received from Alexa but the problem is that this command will never be executed because the shutter relays are on/off lamp for Alexa.
Even configuring a dummy PWM out in the template and making it changed from Alexa the shutter is not moved because obviously the device id received in the hue API command will not match the shutter id and the function ShutterState returns FALSE.
So i'm wondering how to reach it. Is these code really never executable or i'm missing something?
Thanks

**Comment by ascillato** [comment]
Hi, as you didn't have any answer here, you can try addressing this to the [Tasmota Support Chat](https://discord.gg/Ks2Kzd4).

**Comment by oponyx** [comment]
Hi,
no answers received on the Tasmota Support Chat, so i wanted to modify the code to use one PWM out per Shutter to control it through the hue emulation and Alexa.
The code translates the device id received from the hue API and routes the command to the shutters.
![console](https://user-images.githubusercontent.com/78806035/141613841-1cde7ebe-2aad-4da2-9cb5-a15cc5d34c3a.jpg)
As you can see in the console capture the shutters receive the command now
To have this working you need:
* SetOption80 and SetOption68 enabled,
* hide the friendly names of the shutters relays to Alexa 
* For every shutter configured add a dummy PWM out on a free pin in the template of your Tasmota device. 
Then make Alexa discover your PWM lamp and use it to drive the shutter!

beware: I've only tested it with max two shutters and only in ShutterMode1/2. I don't have the right hardware to test all the modes...
Link to the modified fork: https://github.com/oponyx/Tasmota/tree/dummy_pwm

Regards

**Reply by ramos45** [reply]
Sorry, i'm a newbie. How can i flash It?

**Reply by ramos45** [reply]
> Hi, no answers received on the Tasmota Support Chat, so i wanted to modify the code to use one PWM out per Shutter to control it through the hue emulation and Alexa. The code translates the device id received from the hue API and routes the command to the shutters. ![console](https://user-images.githubusercontent.com/78806035/141613841-1cde7ebe-2aad-4da2-9cb5-a15cc5d34c3a.jpg) As you can see in the console capture the shutters receive the command now To have this working you need:
> 
>     * SetOption80 and SetOption68 enabled,
> 
>     * hide the friendly names of the shutters relays to Alexa
> 
>     * For every shutter configured add a dummy PWM out on a free pin in the template of your Tasmota device.
>       Then make Alexa discover your PWM lamp and use it to drive the shutter!
> 
> 
> beware: I've only tested it with max two shutters and only in ShutterMode1/2. I don't have the right hardware to test all the modes... Link to the modified fork: https://github.com/oponyx/Tasmota/tree/dummy_pwm
> 
> Regards


Sorry, i'm a newbie. How can i flash It?

**Reply by oponyx** [reply]
Hi,
you have to build the firmware for your device, if you don't know how to do it please let me know your hardware, i'll try to help you...

**Reply by ramos45** [reply]
I have a wemos d1 mini. If i can test It and works fine for me, i use some wemos D1 mini and buy a esp32 for a room that i need more gpio pins

**Reply by oponyx** [reply]
[tasmota.bin.gz](https://github.com/arendst/Tasmota/files/9843719/tasmota.bin.gz)
This firmware should be good for the wemos d1 mini, flash it using Tasmotizer

**Reply by ramos45** [reply]
Thanks!

**Reply by stefanbode** [reply]
Shuttersupport through HUE (Lights) is a tricky beast. I tried my best to link it to a dimable light but it does not work really good. Also OPEN/CLOSE is not recognized with lights. At the end my suggestion it to define a shutter in HA oder openHUB and do the Alexa integration in one these automation software suites

**Reply by oponyx** [reply]
> Shuttersupport through HUE (Lights) is a tricky beast. 

Oh yes I've seen... 
@stefanbode you made a great piece of code, it's wonderful!
I agree using HA or openHAB is the best way to make it work, but I hadn't the hardware and wanted to try to find a way to link PWM lamp to shutters...
I digged inside the code and the only method to route the command that i've found(or at least the fastest one) has been to change the the device id of the dimmable Lamp on run time.
